internal/controller/object/informers.go

```diff
@@ -0,0 +1,273 @@
+			continue
+		}
+
+		// don't forget to call cancelFn in error cases to avoid leaks. In the
+		// happy case it's called from the go routine starting the cache below.
+		ctx, cancelFn := context.WithCancel(context.Background())
+
+		u := kunstructured.Unstructured{}
+		u.SetGroupVersionKind(gvk)
+		inf, err := ca.GetInformer(ctx, &u, cache.BlockUntilSynced(false)) // don't block. We wait in the go routine below.
+		if err != nil {
+			cancelFn()
+			log.Debug("failed getting informer", "error", err)
+			continue
+		}
+
+		if _, err := inf.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				ev := runtimeevent.GenericEvent{
+					Object: obj.(client.Object),
+				}
+
+				i.sink(providerConfig, ev)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				old := oldObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				obj := newObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				if old.GetResourceVersion() == obj.GetResourceVersion() {
+					return
+				}
+
+				ev := runtimeevent.GenericEvent{
+					Object: obj,
+				}
+
+				i.sink(providerConfig, ev)
+			},
+			DeleteFunc: func(obj interface{}) {
+				ev := runtimeevent.GenericEvent{
+					Object: obj.(client.Object),
+				}
+
+				i.sink(providerConfig, ev)
+			},
+		}); err != nil {
+			cancelFn()
+			log.Debug("failed adding event handler", "error", err)
+			continue
+		}
+
+		go func() {
+			defer cancelFn()
+
+			log.Info("Starting resource watch")
+			_ = ca.Start(ctx)
+		}()
+
+		i.lock.Lock()
+		i.resourceCaches[gvkWithConfig{providerConfig: providerConfig, gvk: gvk}] = resourceCache{
+			cache:    ca,
+			cancelFn: cancelFn,
+		}
+		i.lock.Unlock()
+
+		// wait for in the background, and only when synced add to the routed cache
+		go func() {
+			if synced := ca.WaitForCacheSync(ctx); synced {
+				log.Debug("Resource cache synced")
+			}
+		}()
+	}
+}
+
+func (i *resourceInformers) StopWatchingResources(ctx context.Context, providerConfig string, gvks ...schema.GroupVersionKind) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	for _, gvk := range gvks {
+		ca, found := i.resourceCaches[gvkWithConfig{providerConfig: providerConfig, gvk: gvk}]
+		if !found {
+			continue
+		}
+		// Check if there are any other objects referencing this resource GVK.
+		list := v1alpha2.ObjectList{}
+		if err := i.objectsCache.List(ctx, &list, client.MatchingFields{resourceRefGVKsIndex: refKeyProviderGVK(providerConfig, gvk.Kind, gvk.Group, gvk.Version)}); err != nil {
+			i.log.Debug("cannot list objects referencing a certain resource GVK", "error", err, "fieldSelector", resourceRefGVKsIndex+"="+refKeyProviderGVK(providerConfig, gvk.Kind, gvk.Group, gvk.Version))
+		}
+
+		inUse := false
+		for _, o := range list.Items {
+			// We only care about objects that are not being deleted. Otherwise,
+			// we are getting into deadlocks while stopping the watches during
+			// deletion.
```

Comment:
I don't understand this. We should keep watching deleting objects. Why not?

Reply:
This is a little tricky and I couldn't find a better solution. 

Let me try to explain with an example; let's say I have `Object A` creating a `ConfigMap` on a remote cluster. When the `Object A` is deleted, I want to stop the relevant cache as soon as possible. If I leave it to the garbage collector, the cache throws errors until the next GC run after the remote cluster is gone (which is a typical provider-k8s). So, in addition to GC, I introduced this `StopWatchingResources` method which is invoked when `Object A` is deleted.
However, the problem is when I list Objects watching ConfigMaps on that remote cluster to see if no others left, I am also getting `Object A` since it is still around (until we remove its finalizer). I had two options at this moment, I would either
a. ignore Objects that are being deleted (hence will go away soon and do not need to watch) 
b. ignore the calling Object, i.e. `Object A`.  

Can't remember exactly why I went with (a) since (b) makes more sense now. But, I vaguely remember some issue with (b) which I'll double check 🤔

Reply:
Ok, I think I remembered. When I use (b), if multiple resources using the same cache being deleted at the same time, none of the `StopWatchingResources` calls will do the clean up since the other (being deleting) objects are still around. This is not a problem with (a).

Reply:
Note the difference between "being deleted", and "having deletionTimestamp set". You will see events for either case, but should only stop on the first. The second case is a normal part in the life-cycle, and can take long time with finalizers.

Reply:
Wondering, we get deletion events when the object are really gone. Could we somehow **cheaply** check whether there are objects left for that provider and GVK? Can the indexes help?

Reply:
As we discussed offline, I removed that "immediate watch stopping logic" at the cost of having some annoying logs until garbage collection kicks off. We can check for possible solutions later depending on how annoying they are :D